text/0000-conditional-compilation-checking.md

```diff
@@ -0,0 +1,585 @@
+> shortened to > `--check-cfg names()` in order to enable checking condition names.
+
+## Drawbacks
+
+* Adds complexity, in the form of additional command-line options. Fortunately, this is
+  complexity that will be mainly be exposed to build systems, such as Cargo.
+* As with all lints, correct code may be trigger lints. Developers will need to take time to
+  examine them and see whether they are legitimate or not.
+* To take full advantage of this, build systems (including but not limited to Cargo) must be
+  updated. However, for those systems that are not updated, there is no penalty or drawback,
+  since `--check-cfg` is opt-in.
+
+* This lint will not be able to detect invalid `#[cfg]` tests that are within modules that
+  are not compiled, presumably because an ancestor `mod` is disabled due to a.  For example:
+
+  File `lib.rs` (root module):
+  ```rust
+  #[cfg(feature = "this_is_disabled_but_valid")]
+  mod foo
+  ```
+
+  File `foo.rs` (nested module):
+  ```rust
+  #[cfg(feature = "oooooops_this_feature_is_misspelled_and_invalid")]
+  mod uh_uh;
+  ```
+
+  The invalid `#[cfg]` attribute in `foo.rs` will not be detected, because `foo.rs` was not
+  read and parsed. This is a minor drawback, and should not prevent users from benefitting
+  from checking in most common situations.
+
+## Rationale and alternatives
+
+This design enables checking for a class of bugs at compile time, rather than detecting them by 
+running code.
+
+This design does not break any existing usage of Rustc. It does not change the meaning of existing
+programs or existing Rustc command-line options. It is strictly opt-in. If the verification that
+this feature provides is valuable, then it could be promoted to a warning in the future, or
+eventually an error. There would need to be a cleanup period, though, where we detected failures in
+existing crates and fixed them.
+
+The impact of not doing this is that a class of bugs may go undetected. These bugs are often easy
+to find in relatively small systems of code, but experience shows that these kinds of bugs are much
+harder to verify in large code bases. Rust should enable developers to scale up from small to large
+systems, without losing agility or reliability.
+
+## Prior art
+
+Rust has a very strong focus on finding defects at compile time, rather than allowing defects to be
+detected much later in the development cycle. Statically checking that conditional compilation is
+used correctly is consistent with this approach.
+
+Many languages have similar facilities for conditional compilation. C, C++, C#, and many of their
+variants make extensive use of conditional compilation. The author is unaware of any effort to
+systematically verify the correct usage of conditional compilation in these languages.
+
+## Unresolved questions
+
+During the RFC process, I expect to resolve the question of what the exact rustc command-line

```

Comment:
nit: this section is slightly outdated i think

Reply:
Agreed; I've updated it.